lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,743 @@
+		credentials => [
+			{
+				type => 'password',
+				temporary => $JSON::PP::false,
+				value => $password
+			}
+		],
+		attributes => [
+			name => [$user_ref->{name}],
+			locale => [$user_ref->{initial_lc}],
+			country => [$user_ref->{initial_cc}],
+		]
+	};
+	my $json = encode_json($api_request_ref);
+
+	my $create_user_request = HTTP::Request->new(POST => $keycloak_users_endpoint);
+	$create_user_request->header('Content-Type' => 'application/json');
+	$create_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	$create_user_request->content($json);
+	my $new_user_response = LWP::UserAgent::Plugin->new->request($create_user_request);
+	unless ($new_user_response->is_success) {
+		display_error_and_exit($new_user_response->content, 500);
+	}
+
+	my $get_user_request = HTTP::Request->new(GET => $new_user_response->header('location'));
+	$get_user_request->header('Content-Type' => 'application/json');
+	$get_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	my $get_user_response = LWP::UserAgent::Plugin->new->request($get_user_request);
+	unless ($get_user_response->is_success) {
+		display_error_and_exit($get_user_response->content, 500);
+	}
+
+	my $json_response = $get_user_response->decoded_content(charset => 'UTF-8');
+	my @created_users = decode_json($json_response);
+	return $created_users[0];
+}
+
+sub create_user_in_product_opener ($id_token) {
+	unless ($id_token) {
+		return;
+	}
+
+	my $user_ref = {};
+	$user_ref->{email} = $id_token->{'email'};
+	$user_ref->{userid} = $id_token->{'preferred_username'};
+	$user_ref->{name} = $id_token->{'name'};
+
+	my $user_id = $user_ref->{userid};
+	my $user_file = "$BASE_DIRS{USERS}/" . get_string_id_for_lang("no_language", $user_id) . ".sto";
+	store($user_file, $user_ref);
+
+	# Store email
+	my $emails_ref = retrieve("$BASE_DIRS{USERS}/users_emails.sto");
+	my $email = $user_ref->{email};
+
+	if ((defined $email) and ($email =~ /\@/)) {
+		$emails_ref->{$email} = [$user_id];
+	}
+
+	if (defined $user_ref->{old_email}) {
+		delete $emails_ref->{$user_ref->{old_email}};
+		delete $user_ref->{old_email};
+	}
+
+	store("$BASE_DIRS{USERS}/users_emails.sto", $emails_ref);
+	return $user_id;
+}
+
+=head2 get_token_using_password_credentials($username, $password)
+
+Gets a token for the user.
+
+Method uses the Resource Owner Password Credentials Grant to
+with the given credentials, and pre-configured Client ID,
+and Client Secret.
+
+=head3 Arguments
+
+=head4 Name of the user $usersname
+
+=head4 Password given at sign-in $password
+
+=head3 Return values
+
+Open ID Access token, or undefined if sign-in wasn't successful.
+
+=cut
+
+sub get_token_using_password_credentials ($username, $password) {
+	_ensure_oidc_is_discovered();
+
+	my $token_request = HTTP::Request->new(POST => $oidc_discover_document->{token_endpoint});
+	$token_request->header('Content-Type' => 'application/x-www-form-urlencoded');
```

Comment:
Couldn't we use json instead for clarity ?

Reply:
I think form-encoded data is usually required by OAuth2 token endpoints